addons/gamification/models/gamification_karma_tracking.py

```diff
@@ -0,0 +1,100 @@
+    _name = 'gamification.karma.tracking'
+    _description = 'Track Karma Changes'
+
+    user_id = fields.Many2one('res.users', 'User', required=True)
+    old_value = fields.Integer('Old Karma Value', required=True, readonly=1)
+    new_value = fields.Integer('New Karma Value', required=True, readonly=1)
+    consolidation_type = fields.Selection(selection=[('none', 'Not Consolidated'),
+                                                     ('daily', 'Daily Consolidated'),
+                                                     ('monthly', 'Monthly Consolidated'),
+                                                     ('yearly', 'Yearly Consolidated')],
+                                          default='none')
+    tracking_date = fields.Datetime(default=fields.Datetime.now)
+
+    @api.model
+    def _consolidate_last_day(self):
+        to_date = datetime.now()
+        from_date = to_date - relativedelta(days=1)
+        self._process_consolidate(from_date, to_date, 'none', 'daily', '%Y-%m-%d')
+
+    @api.model
+    def _consolidate_last_month(self):
+        """ When consolidating daily values in monthly values we need to do it for the month before the last one
+            so that we will keep at least 30 daily values. If we immediately consolidate all daily values we won't
+            be able to retrieve last week karma changes accurately.
+        """
+        to_date = datetime.now() - relativedelta(months=1)
+        from_date = to_date - relativedelta(months=1)
+        self._process_consolidate(from_date, to_date, 'daily', 'monthly', '%Y-%m-01')
+
+    @api.model
+    def _consolidate_last_year(self):
+        """ Like the _consolidate_last_month we consolidate the monthly values in yearly values for the year before
+            the last one.
+        """
+        to_date = datetime.now() - relativedelta(years=1)
+        from_date = to_date - relativedelta(years=1)
+        self._process_consolidate(from_date, to_date, 'monthly', 'yearly', '%Y-01-01')
+
+    def _process_consolidate(self, from_date, to_date, old_consolidation_type, new_consolidation_type, tracking_date_format):
+        """
+        Consolidate all old_consolidation_type records which tracking_date is between from_date to to_date
+        into one new_consolidation_type record with a new tracking_date
+
+        :param from_date: the date from which the method will consolidate the records (included)
+        :param to_date: the date until which the method will consolidate the records (excluded)
+        :param old_consolidation_type: the consolidation_type of the records to consolidate
+        :param new_consolidation_type: the consolidation_type of the resulting record
+        :param tracking_date_format: the date format that will be used for the tracking_date of the resulting record
+        :return:
+        """
+        select_query = """
+                        SELECT user_id, MIN(old_value) as old_value, MAX(new_value) as new_value
```

Comment:
Small question: is someone loose karma, we should not really take the max but the last value, isn't it ? It seems consolidation should be first entry in month -> last entry in month. If no entry, no consolidation, obviously.